homeassistant/components/scheduler/schedule.py

```
@@ -0,0 +1,378 @@
+            if not start:
+                return (None, None)
+
+            if self.end_datetime:
+                end = start + self.instance_duration
+            else:
+                end = None
+
+            return (start, end)
+
+        return (None, None)
+
+    def as_dict(self) -> dict:
+        """Return the schedule as a dict."""
+        the_dict = {
+            CONF_ENTITY_ID: self.entity_id,
+            CONF_START_DATETIME: self.start_datetime.isoformat(),
+        }
+
+        if self.end_datetime:
+            the_dict[CONF_END_DATETIME] = self.end_datetime.isoformat()
+        else:
+            the_dict[CONF_END_DATETIME] = None
+
+        if self.recurrence:
+            the_dict[CONF_RECURRENCE] = str(self.recurrence)
+        else:
+            the_dict[CONF_RECURRENCE] = None
+
+        return the_dict
+
+    async def async_revert(self, instance: ScheduleInstance) -> None:
+        """Revert the instance."""
+        await instance.async_revert()
+        await self._async_schedule()
+
+    async def async_trigger(self, instance: ScheduleInstance) -> None:
+        """Trigger the instance."""
+        await instance.async_trigger()
+        if not self.instance_duration:
+            await self._async_schedule()
+
+    @callback
+    def async_turn_off(self) -> None:
+        """Disable the schedule."""
+        if self.active_instance:
+            self.active_instance.async_cancel()
+            self.active_instance = None
+        if self._async_revert_listener:
+            self._async_revert_listener()
+            self._async_revert_listener = None
+        if self._async_trigger_listener:
+            self._async_trigger_listener()
+            self._async_trigger_listener = None
+
+        self._is_on = False
+
+    @callback
+    def async_turn_on(self) -> None:
+        """Enable the schedule."""
+        if self.expired:
+            _LOGGER.error("Refusing to turn on expired schedule: %s", self)
```

Let's raise instead of swallowing the error. It should not be called. `raise ValueError("Cannot turn on expired schedule")`